File: src/main/java/com/durhack/sharpshot/nodes/Container.java
package com.durhack.sharpshot.nodes;

import com.durhack.sharpshot.Bullet;
import com.durhack.sharpshot.Coordinate;
import com.durhack.sharpshot.Direction;
import com.durhack.sharpshot.INode;
import javafx.scene.Node;
import javafx.scene.paint.Color;
import javafx.scene.shape.Rectangle;
import javafx.util.Pair;
import org.jetbrains.annotations.NotNull;

import java.math.BigInteger;
import java.util.*;

//TODO how do we deal with the container returning multiple outputs over multiple ticks?

public class Container implements INode {
    private int width;
    private int height;

    private Direction rotation = Direction.UP;

    @NotNull
    private Map<Coordinate, INode> nodes = new HashMap<>();

    @NotNull
    private Map<Coordinate, Bullet> bullets = new HashMap<>();

    public Container(int width, int height) {
        this.width = width;
        this.height = height;
    }

    public int getHeight() {
        return height;
    }

    public boolean noBullets() {
        return bullets.isEmpty();
    }

    @Override
    public @NotNull Direction getRotation() {
        return rotation;
    }

    @Override
    public void rotateClockwise() {
        rotation = Direction.clockwiseOf(rotation);
    }

    @NotNull
    @Override
    public Map<Direction, BigInteger> run(@NotNull Bullet bullet) {
        return new HashMap<>();
    }

    @NotNull
    public Map<Coordinate, INode> getNodes() {
        return nodes;
    }

    @NotNull
    public Map<Coordinate, Bullet> getBullets() {
        return bullets;
    }

    public void start(List<BigInteger> input) {
        // In nodes add input
        Map<Coordinate, Bullet> newBullets = new HashMap<>();
        for (int i = 0; i < input.size(); i++) {
            for (Coordinate coordinate : getNodes().keySet()) {
                INode x = getNodes().get(coordinate);
                if (x instanceof NodeIn && ((NodeIn) x).getIndex() == i) {
                    Map<Direction, BigInteger> bulletParams = ((NodeIn) x).into(input.get(i));

                    for (Map.Entry<Direction, BigInteger> newBulletEntry : bulletParams.entrySet()) {
                        Bullet newBullet = new Bullet(x.getRotation(), newBulletEntry.getValue());
                        Coordinate newCoordinate = coordinate.plus(newBullet.getDirection());
                        newBullets.put(newCoordinate, newBullet);
                    }
                }
            }
        }

        bullets.putAll(newBullets);
    }

    /**
     * Bullets on top of node -> store
     * Other bullets -> tick
     * stored bullets -> process and output
     * all bullets -> check
     */
    public void tick() {
        List<Pair<Movement, Bullet>> movements = new ArrayList<>();

        //Bullets on nodes
        Map<Coordinate, Bullet> capturedBullets = new HashMap<>(bullets);
        capturedBullets.keySet().retainAll(nodes.keySet());

        //Bullets not on nodes
        Map<Coordinate, Bullet> freeBullets = new HashMap<>(bullets);
        freeBullets.keySet().removeAll(capturedBullets.keySet());

        //Generate movements for free bullets
        for (Map.Entry<Coordinate, Bullet> entry : freeBullets.entrySet()) {
            Bullet bullet = entry.getValue();
            Coordinate coord = entry.getKey();
            Coordinate newCoord = entry.getKey().plus(bullet.getDirection()).wrap(width, height);
            movements.add(new Pair<>(new Movement(coord, newCoord), bullet));
        }

        //Update captured bullets
        for (Map.Entry<Coordinate, Bullet> entry : capturedBullets.entrySet()) {
            Coordinate coordinate = entry.getKey();
            INode node = nodes.get(coordinate);

            //TODO this brings great shame onto my family
            Direction bulletDirection = entry.getValue().getDirection();
            Direction dir = Direction.UP;
            while (dir != node.getRotation()) {
                dir = dir.clockwise();
                bulletDirection = bulletDirection.antiClockwise();
            }
            Bullet bullet = new Bullet(bulletDirection, entry.getValue().getValue());

            Map<Direction, BigInteger> bulletParams = node.run(bullet);

            for (Map.Entry<Direction, BigInteger> newBulletEntry : bulletParams.entrySet()) {
                //TODO this too
                bulletDirection = newBulletEntry.getKey();
                dir = Direction.UP;
                while (dir != node.getRotation()) {
                    dir = dir.clockwise();
                    bulletDirection = bulletDirection.clockwise();
                }

                Bullet newBullet = new Bullet(bulletDirection, newBulletEntry.getValue());
                Coordinate newCoordinate = coordinate.plus(newBullet.getDirection()).wrap(width, height);
                Movement movement = new Movement(coordinate, newCoordinate);
                movements.add(new Pair<>(movement, newBullet));
            }
        }

        //Remove swapping bullets
        List<Movement> read = new ArrayList<>();
        Set<Coordinate> toDelete = new HashSet<>();
        for (Pair<Movement, Bullet> pair : movements) {
            Movement movement = pair.getKey();
            Coordinate from = movement.getFrom();
            Coordinate to = movement.getTo();
            boolean foundSwaps = read.stream().anyMatch(other -> from.equals(other.getTo()) && to.equals(other.getFrom()));
            if(foundSwaps){
                toDelete.add(from);
                toDelete.add(to);
            }
            read.add(movement);
        }
        movements.removeIf(pair -> toDelete.contains(pair.getKey().getFrom()) || toDelete.contains(pair.getKey().getTo()));

        //Remove bullets that end in the same place
        read.clear();
        toDelete.clear();
        for (Pair<Movement, Bullet> pair: movements) {
            Movement movement = pair.getKey();
            Coordinate to = movement.getTo();
            boolean foundSameFinal = read.stream().anyMatch(other -> to.equals(other.getTo()));
            if(foundSameFinal){
                toDelete.add(to);
            }
            read.add(movement);
        }
        movements.removeIf(pair -> toDelete.contains(pair.getKey().getTo()));

        //Move bullets
        Map<Coordinate, Bullet> newBullets = new HashMap<>();
        for (Pair<Movement, Bullet> pair: movements) {
            Movement movement = pair.getKey();
            Bullet bullet = pair.getValue();
            newBullets.put(movement.getTo(), bullet);
        }
        bullets.clear();
        bullets.putAll(newBullets);
    }

    public int getWidth() {
        return width;
    }

    @Override
    public @NotNull Node toGraphic() {
        return new Rectangle(32.0, 32.0, Color.GREEN);
    }


    public void reset() {
        bullets.clear();
        for (INode n : getNodes().values())
            n.reset();
    }
}

File: src/main/java/com/durhack/sharpshot/nodes/Container.kt
package com.durhack.sharpshot.nodes

import com.durhack.sharpshot.Bullet
import com.durhack.sharpshot.Coordinate
import com.durhack.sharpshot.Direction
import com.durhack.sharpshot.INode
import com.durhack.sharpshot.nodes.io.AbstractInputNode
import com.durhack.sharpshot.nodes.io.InNode
import com.durhack.sharpshot.util.Listeners
import com.durhack.sharpshot.util.Movement

import java.math.BigInteger
import java.util.*

class Container(val width: Int, val height: Int) {
    val completionListeners = Listeners()

    val nodes = HashMap<Coordinate, INode>()
    val bullets = HashMap<Coordinate, Bullet>()

    fun start(input: List<BigInteger>) {
        val newBullets = HashMap<Coordinate, Bullet>()

        // Input nodes spawn 0 if their index == 0
        for (coordinate in nodes.keys) {
            val node = nodes[coordinate]
            if (node is AbstractInputNode) {
                val bulletParams = node.input(input)

                for ((_, value) in bulletParams) {
                    val newBullet = Bullet(node.rotation, value)
                    val newCoordinate = coordinate.plus(newBullet.direction)
                    newBullets[newCoordinate] = newBullet
                }
            }
        }

        bullets.putAll(newBullets)
    }

    fun firstAvailableInputIndex(): Int {
        val inputs = nodes.values
                .filter { node -> node is InNode }
                .map { node -> (node as InNode).index }

        var counter = 1
        while (inputs.contains(counter)) {
            counter++
        }
        return counter
    }

    /**
     * Bullets on top of node -> store
     * Other bullets -> tick
     * stored bullets -> process and output
     * all bullets -> check
     */
    fun tick() {
        var haltNodeHit = false

        val bulletMovements = mutableListOf<Pair<Movement, Bullet>>()

        //Bullets on nodes
        val captured = bullets.mapNotNull { (coord, bullet) ->
            val node = nodes[coord] ?: return@mapNotNull null
            return@mapNotNull Triple(coord, node, bullet)
        }.toList()


        //Bullets not on nodes
        val capturedBullets = captured.map { it.third }.toSet()
        val freeBullets = bullets.filterValues { it !in capturedBullets }

        //Generate movements for free bullets
        bulletMovements.addAll(freeBullets.map {(coord, bullet) ->
            val newCoord = coord.plus(bullet.direction).wrap(width, height)
            return@map Movement(coord, newCoord) to bullet
        })

        //Update captured bullets
        captured.forEach {(coord, node, bullet) ->
            // special case
            // if any halt nodes get hit by a bullet, halt at end of
            if (!haltNodeHit && node is HaltNode) {
                haltNodeHit = true
            }

            //TODO this brings great shame onto my family
            var bulletDirection = bullet.direction
            var dir = Direction.UP
            while (dir != node.rotation) {
                dir = dir.clockwise()
                bulletDirection = bulletDirection.antiClockwise()
            }
            val rotatedBullet = Bullet(bulletDirection, bullet.value)

            node.run(rotatedBullet).forEach { direction, value ->
                //TODO this too
                bulletDirection = direction
                dir = Direction.UP
                while (dir != node.rotation) {
                    dir = dir.clockwise()
                    bulletDirection = bulletDirection.clockwise()
                }

                val newBullet = Bullet(bulletDirection, value)
                val newCoordinate = coord.plus(newBullet.direction).wrap(width, height)
                val movement = Movement(coord, newCoordinate)
                bulletMovements.add(movement to newBullet)
            }
        }

        //Remove swapping bullets
        //When two bullets are adjacent and each is trying to move into the space the other is occupying, a naive solution
        //would allow them to swap without colliding.
        //This solves that problem
        val swapProcessed = mutableSetOf<Movement>()
        val swapDelete = mutableSetOf<Coordinate>()
        bulletMovements.forEach { (movement, _) ->
            val from = movement.from
            val to = movement.to
            val foundSwaps = swapProcessed.stream().anyMatch { other -> from == other.to && to == other.from }
            if (foundSwaps) {
                swapDelete.add(from)
                swapDelete.add(to)
            }
            swapProcessed.add(movement)
        }
        bulletMovements.removeIf { (movement, _) -> swapDelete.contains(movement.from) || swapDelete.contains(movement.to) }

        //Remove bullets that end in the same place
        val collideProcessed = mutableSetOf<Movement>()
        val collideDelete = mutableSetOf<Coordinate>()
        bulletMovements.forEach { (movement, _) ->
            val to = movement.to
            val foundSameFinal = collideProcessed.stream().anyMatch { other -> to == other.to }
            if (foundSameFinal) {
                collideDelete.add(to)
            }
            collideProcessed.add(movement)
        }
        bulletMovements.removeIf { (movement, _) -> collideDelete.contains(movement.to) }

        //Move bullets
        val newBullets = bulletMovements.map { (movement, bullet) ->
            movement.to to bullet
        }.toMap()

        bullets.clear()
        bullets.putAll(newBullets)

        if (haltNodeHit) {
            halt()
        }
    }

    private fun halt() {
        completionListeners()
        reset()
    }

    fun reset() {
        bullets.clear()
        nodes.forEach { _, node -> node.reset() }
    }

    fun clearAll() {
        bullets.clear()
        nodes.clear()
    }
}